DDF-3997 Create a service to support an incoming query and outbound format for generating files

Corey-Collins opened the pull request:
<!--
Uncomment this block and include PR # of change against 2.13.x where appropriate.
When doing so, it is reasonable to delete much of the rest of the PR description, leaving only:
- This uncommented block with link to 2.13.x PR
- Reviewers tagged
- Teams tagged
- Committers tagged
- The link to the "Notes on Review Process"

##### ABBREVIATED REVIEW BETWEEN 2.13.X AND MASTER IS IN EFFECT
Link to 2.13.x PR: #XXXX
____
-->

#### What does this PR do?
- Refactors much of the existing QueryApplication to support handler classes, allowing for unit testing on Spark endpoints.
- Creates a `/internal/cql/transform/{transformerId}` endpoint used to export a file from a result set.
- The result set from the query response will be transformed to the matching id, and returned in the http response in the file format that matches the transformer's mime-type.
- Optional arguments can be passed to the transformer from the http request body using an `arguments` json key.

#### Who is reviewing it? 
<!--(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)-->
@brianfelix 
@leo-sakh 
@bellcc
@jrnorth 
#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Ask 2 committers to review/merge the PR and tag them here.
<!--
If you don't know who to ask, you can request reviews in https://groups.google.com/forum/#!forum/ddf-developers .
(please choose ONLY two committers from below, delete the rest)
-->
@millerw8 
@bdeining 
#### How should this be tested?
<!--(List steps with links to updated documentation)-->

- (Optional) Disable web sockets in Admin > System > Catalog UI Search
- Populate local catalog with metacard(s) (via postman or ui uploader)
- Search using a transformer id (I recommend `/cql/transform/xml`). You can do this in postman, however I was able to do this easier with the `Requestly` chrome extension, where you replace the host in a request from `cql` to `cql/transform/xml`, and execute that query from the run button in DDF ui (and view the response data in network dev tools).
- Example request body: (src and cql required keys)
```
{
    "src":"cache",
    "start":1,
    "count":250,
    "cql":"((anyText ILIKE '%') AND ((\"metacard_source\" = 'ddf.distribution')))",
    "sorts": [
        {
            "attribute":"modified",
            "direction":"descending"
            
        }
    ],
    "id":"6d9ed55a-1b6c-4b94-992f-c4e7374ae3dc"
}
```
- In network dev tools or postman ensure the response header returns `attachment;filename=export-{time}{fileformat}` (for ex: `attachment;filename=export-2018-07-30T20:36:15.302Z.xml`) and contains the metacard(s) data in proper transformation data format (in my case xml).

#### Any background context you want to provide?

Future tickets would like a result set to be exported to various file formats. Transformers will need to be made for these formats, and this service will allow the UI to perform this operation to access those transformers.

#### What are the relevant tickets?
[DDF-3997](https://codice.atlassian.net/browse/DDF-3997)
#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/codice/ddf?pullRequest=3516) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/codice/ddf?pullRequest=3516) before we can accept your contribution.<br/><hr/>**Corey Collins** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/codice/ddf?pullRequest=3516) it.</sub>

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -98,9 +122,77 @@ public void init() {
+            MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
+            String fileExt = allTypes.forName(mimeType).getExtension();
+
+            res.status(200);
```
✏️ `Response.OK`

rzwiefel replied:
Probably shouldn't be using jaxrs within a spark endpoint.

Corey-Collins replied:
I don't believe this is included in DDF anyway

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -98,9 +122,77 @@ public void init() {
+          LOGGER.debug(
+              String.format(
+                  "Could not find transformer id %s to match cql request.", transformerId));
+          res.status(404);
```
✏️ `Response.NOT_FOUND`

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -98,9 +122,77 @@ public void init() {
+
+            String id = (String) queryResponseTransformer.getProperty("id");
+
+            if (!id.equals(transformerId)) {
```
✏️ This is more readable if you change this condition so that when it evaluates as true it performs the transformation

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -98,9 +122,77 @@ public void init() {
+                StringUtils.isNotBlank(acceptEncoding)
+                    && acceptEncoding.toLowerCase().contains("gzip");
+
+            if (!shouldGzip) {
```
❓ Is this header definitely needed? And if it is should we do more than just log the error?

Corey-Collins replied:
it's for compression, and isn't absolutely needed.

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -98,9 +122,77 @@ public void init() {
+              }
+            }
+
+            LOGGER.trace(String.format("Response sent in %s file format.", fileExt));
```
✏️ You can use parameterized logging here like this...

```
String str = "World";
LOGGER.trace("Hello {}", str);
```

Corey-Collins replied:
I had originally meant to do that but I've been getting `[FormatString] extra format arguments`

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -98,9 +122,77 @@ public void init() {
+
+            LOGGER.trace(String.format("Response sent in %s file format.", fileExt));
+
+            return "";
```
❓ Why return the empty string?

Corey-Collins replied:
the return value is what is included in the body, in the case an attachment is made the body is empty

Corey-Collins replied:
as @rzwiefel mentioned it is odd to have the body include return data in one case but as an attachment in another. but that is also what the team recommended, I would like to hear some input on that. I don't believe the results can be returned in the body instead of an attachment since the attachment is the data that the UI download as an external file.

Corey-Collins commented:
@rzwiefel @bellcc I changed the endpoint to `/internal/cql/transform/{transformerId}` so that it is its own separate endpoint from `/internal/cql`. I addressed most of Chris' comments in this commit as well.

rzwiefel commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -76,6 +88,12 @@
 
   private FeatureService featureService;
 
+  private QueryResponse queryResponse;
```
❗️ we should not be using a field on the class for storing the queryResponse as that could lead to issues.

Instead we can use the returned response in the same scope (line 142)

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,75 @@ public void init() {
+                      .transform(this.queryResponse, new HashMap<>());
+
+              String mimeType = (String) queryResponseTransformer.getProperty("mime-type");
+              MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
```
Could use the `MimeTypeMapper` service to get the extension.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,75 @@ public void init() {
+
+              res.status(200);
+
+              String acceptEncoding = req.headers("Accept-Encoding");
```
[HttpHeaders](https://docs.oracle.com/javaee/7/api/javax/ws/rs/core/HttpHeaders.html) has constants for these.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,75 @@ public void init() {
+                  StringUtils.isNotBlank(acceptEncoding)
+                      && acceptEncoding.toLowerCase().contains("gzip");
+
+              if (!shouldGzip) {
```
✏️ I'd remove this block and instead log that the requester does accept gzip in the `if (shouldGzip)` block below.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,75 @@ public void init() {
+              res.header("Content-Disposition", attachment);
+
+              if (shouldGzip) {
+                res.raw().addHeader("Content-Encoding", "gzip");
```
`res.header()`

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +132,75 @@ public void init() {
+
+              try (OutputStream servletOutputStream = res.raw().getOutputStream();
+                  InputStream resultStream = content.getInputStream()) {
+                if (shouldGzip) {
```
According to the [Spark documentation](http://sparkjava.com/documentation#gzip), to GZIP the response you only have to add the header and it will be done automatically.

Corey-Collins replied:
Removing the GZIP OutputStream (leaving just what's in the else block) causes an net::ERR_CONTENT_DECODING_FAILED response

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,72 @@ public void init() {
 
+    post(
+        "/cql/transform/:transformerId",
+        (req, res) -> {
```
❗️ this lambda is doing a lot of things. It parses the request parameters, it executes the cql query, it constructs the response (as it goes). Let's try to extract the functionality herein into separate methods each having a single responsibility, please 😄

bdeining replied:
This unfortunately has been the convention in all of the endpoints written with the Spark application.  While I agree with the separation of concerns suggestion here, it might be best to do this in a separate ticket, as this may ultimately be out of scope with the task at hand.

leo-sakh replied:
While that's understandable, I would suggest we go ahead and refactor this as a starting point since it's fairly small. Otherwise we're just piling on tech debt (assuming there's not already a ticket to refactor this in it's entirety).

Corey-Collins replied:
now that i've broken down the lambda function a lot it shouldn't be difficult to translate it to a regular function

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,72 @@ public void init() {
+
+            if (id.equals(transformerId)) {
+              BinaryContent content =
+                  bundleContext
```
❓ I'm not 100% on this, but I think we can actually just get a list of service references based on a filter that filters them to only be what we're looking for. `BundleContext` is a little broad in my opinion and if we could avoid it that would be good.

bdeining replied:
Since the transformer ID is read off of the incoming request; I don't think filter parameter can be used in blueprint.  This seems to be the best approach to my knowledge to conditionally look up the correct service references.

jrnorth replied:
Well you could use `bundleContext.getServiceReferences(QueryResponseTransformer.class.getName(), "(id=<transformerId>)")` and not have to loop over the list of service references.

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,72 @@ public void init() {
+
+              LOGGER.trace("Response sent in {} file format.", fileExt);
+
+              return "";
```
✏️ a `""` has no semantic meaning. Could you create a string variable with a name that describes this case?

rzwiefel commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlQueryResponse.java:
```diff
@@ -50,6 +51,8 @@
 
   private final Status status;
 
+  private QueryResponse queryResponse;
```
✏️  If this is set in the ctor and never modified, we should be able to make it `final` as well.

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,73 @@ public void init() {
+
+              LOGGER.trace("Response sent in {} file format.", fileExt);
+
+              final String responseBody = "";
```
Is this resp body always suppose to be empty string?

Bdthomson commented on catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -259,6 +259,12 @@
 
+    <bean id="queryResponseTransformer" class="ddf.catalog.util.impl.SortedServiceReferenceList"/>
+    <reference-list interface="ddf.catalog.transform.QueryResponseTransformer"
+                    member-type="service-reference">
```
You will probably want to add a filter on this reference-list for a new property on query response transformers that should be available for table export.

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +133,73 @@ public void init() {
+          CqlRequest cqlRequest = mapper.readValue(body, CqlRequest.class);
+          CqlQueryResponse cqlQueryResponse = executeCqlQuery(cqlRequest);
+
+          for (ServiceReference<QueryResponseTransformer> queryResponseTransformer :
```
❗️ This is actually pretty confusing. At first blush I thought we were iterating over all `QueryResponseTransofrmer`s and collecting the transformed data into a response. Though it appears that only the first match is used since we actually return from inside the `for` loop. Which also means that the code after the loop doesn't get executed if there's a matching transformer? Is this correct?

Corey-Collins replied:
@leo-sakh i could try that but right now im trying out joe's comment that's using `bundleContext.getServiceReferences(QueryResponseTransformer.class.getName(), "(id=<transformerId>)")` not sure which is better

leo-sakh replied:
Might be able to use both...if it makes sense to do so

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -244,6 +300,72 @@ private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
         actionRegistry);
   }
 
+  private void attachFileWithTransformer(
```
not sure if i should break down this helper method even further. for each helper method i may need to pass request, response with it

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +137,34 @@ public void init() {
+
+          LOGGER.trace("Finding transformer to transform query response.");
+
+          ServiceReference<QueryResponseTransformer> queryResponseTransformer =
```
i get an unchecked assignment from ServiceReference to ServiceReference<QueryResponseTransformer>, unsure if this is going to be an issue if left unchecked.

bdeining replied:
I don't think it will be possible for OSGi to return the wrong interface; no an unchecked assignment should be fine in this case

jrnorth replied:
You could change `queryResponseTransformers` to be a `List<ServiceReference<QueryResponseTransformer>>`.

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +137,34 @@ public void init() {
+
+          attachFileWithTransformer(queryResponseTransformer, cqlQueryResponse, req, res);
+
+          return "";
```
im getting comments to both add a variable yet not add a variable for this return statement. i think adding a small comment here would be a nice middle ground for that

Bdthomson commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -244,6 +300,68 @@ private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
+      content =
+          bundleContext
+              .getService(queryResponseTransformer)
+              .transform(cqlQueryResponse.getQueryResponse(), new HashMap<>());
```
You're sending an empty map as the second parameter to this query response transformer. That means you're not sending information about Hidden Fields, Order, or Aliases. 

Please see the existing CsvQueryResponse transformer to see what I mean: 

https://github.com/codice/ddf/blob/master/catalog/transformer/catalog-transformer-csv-queryresponse/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java#L93

Bdthomson commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -244,6 +300,68 @@ private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
+      return;
+    }
+
+    response.status(200);
```
✏️ `HttpStatus.OK_200`

rzwiefel replied:
imo Its fine just using 200, no need to pull in a lib for the number

Bdthomson commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -244,6 +300,68 @@ private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
+    }
+
+    response.status(200);
+    LOGGER.trace("Response sent in {} file format.", fileExt);
```
✏️ You have the queryResponseTransformer in this context so you could add it's `id` to this log statement.

Bdthomson commented on catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -364,6 +370,8 @@
         <property name="featureService" ref="featureService"/>
         <property name="endpointUtil" ref="endpointUtil"/>
+        <argument ref="queryResponseTransformer"/>
+        <argument ref="blueprintBundleContext"/>
```
❓ Where is this `bluerprintBundleContext` coming from?

bdeining replied:
Blueprint provides access to the Bundle Context, which is referenced by `blueprintBundleContext`

Lambeaux commented:
❓ I'm seeing a lot of `try ... ` blocks that just catch and return nothing. How about letting your methods just throw `Exception` and have the errors propagate into registered Spark handlers? It will make the code cleaner.

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -109,6 +137,34 @@ public void init() {
           res.header("Content-Encoding", "gzip");
         });
 
+    post(
```
Any unit tests to update?

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/QueryApplication.java:
```diff
@@ -244,6 +300,55 @@ private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
+
+    response.status(200);
+    LOGGER.trace(
+        "Response sent to transformer id {} in {} file format.",
```
This log message is kind of confusing.

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+
+  private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
+      throws UnsupportedQueryException, SourceUnavailableException, FederationException {
+    QueryRequest request = cqlRequest.createQueryRequest(catalogFramework.getId(), filterBuilder);
```
Could split some of this code up into smaller functions here

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+      }
+    }
+
+    response.status(200);
```
✏️ use constant instead of 200 for status code.

rzwiefel replied:
Honestly for status codes i don't see the need for abstracting it behind a variable. If someone then adds another 200 status code it will be easiest for them to just type 200 than to go find whatever someone has happened to name `200` in this class.

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+    String acceptEncoding = request.headers(HttpHeaders.ACCEPT_ENCODING);
+
+    boolean shouldGzip =
+        StringUtils.isNotBlank(acceptEncoding) && acceptEncoding.toLowerCase().contains("gzip");
```
✏️ make gzip a constant, it's used several times

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+
+  private void attachFileWithTransformer(
+      ServiceReference<QueryResponseTransformer> queryResponseTransformer,
+      CqlQueryResponse cqlQueryResponse,
```
Could split some of this code up into smaller functions here

Corey-Collins replied:
done 👍

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+
+    if (queryResponseTransformer == null) {
+      LOGGER.debug("Could not find transformer with id: {}", transformerId);
+      response.status(404);
```
Use constant for status code

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+
+  @Override
+  public Object handle(Request request, Response response) throws Exception {
+    String transformerId = request.params(":transformerId");
```
Can this be null?

Corey-Collins replied:
transformerId? no it can't

Corey-Collins replied:
if you send `/internal/cql/transform/` it won't reach the endpoint and the user gets `404 not found` so it seems impossible for `transformerId` to be null/empty. But I may handle this for J unit anyway

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+        fileExt);
+  }
+
+  private CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
```
We should find a way to reuse the `executeCqlQuery()` method in `QueryApplication`.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+            .transform(cqlQueryResponse.getQueryResponse(), new HashMap<>());
+
+    String mimeType = (String) queryResponseTransformer.getProperty("mime-type");
+    MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
```
Consider using the `MimeTypeMapper` service or even calling `content.getMimeTypeValue()`.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,245 @@
+
+    response.status(200);
+    LOGGER.trace(
+        "Response sent to transformer id {} in {} file format.",
```
This log message is kind of confusing.

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,193 @@
+
+    try (OutputStream servletOutputStream = response.raw().getOutputStream();
+        InputStream resultStream = content.getInputStream()) {
+      if (request.headers(HttpHeaders.ACCEPT_ENCODING).toLowerCase().contains(GZIP)) {
```
✏️ You could break this condition out into it's own method for readability. It seems like you check for this header in a couple of places too

bellcc commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,193 @@
+    BinaryContent content =
+        bundleContext
+            .getService(queryResponseTransformer)
+            .transform(cqlQueryResponse.getQueryResponse(), new HashMap<>());
```
✏️ You could use `Collections.emptyMap()`

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/util/EndpointUtil.java:
```diff
@@ -392,6 +401,59 @@ public String getJson(Object result) {
 
+  public CqlQueryResponse executeCqlQuery(
+      CqlRequest cqlRequest,
+      CatalogFramework catalogFramework,
```
`EndpointUtil` already has a `CatalogFramework`, and the `FilterAdapter` and `ActionRegistry` should be added to the `EndpointUtil` constructor instead.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,183 @@
+
+  private ActionRegistry actionRegistry;
+
+  private final String GZIP = "gzip";
```
This should be `static` too. (checkstyle would fail this anyway since the name is in all-caps and it is not `static final`)

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,183 @@
+    return "";
+  }
+
+  private void setHttpAttributes(Request request, Response response, BinaryContent content)
```
`setHttpHeaders`?

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,161 @@
+            .findFirst()
+            .orElse(null);
+
+    if (queryResponseTransformer == null) {
```
I wonder if it would be better to check for the transformer before executing the query so we don't spend resources on a query only to throw the results away in the case the transformer doesn't exist.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -259,6 +261,12 @@
         <reference-listener ref="workspaceTransformations" bind-method="bindPlugin" unbind-method="unbindPlugin"/>
     </reference-list>
 
+    <bean id="queryResponseTransformer" class="ddf.catalog.util.impl.SortedServiceReferenceList"/>
```
Should be plural.

jrnorth commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,161 @@
+import spark.Route;
+import spark.utils.IOUtils;
+
+public class CqlTransformHandler implements Route {
```
Needs unit tests.

Corey-Collins replied:
Added unit tests

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,161 @@
+  public Object handle(Request request, Response response) throws Exception {
+    String transformerId = request.params(":transformerId");
+    String body = util.safeGetBody(request);
+    CqlRequest cqlRequest = mapper.readValue(body, CqlRequest.class);
```
In theory, `mapper.readValue` could return null

Corey-Collins replied:
I ran into that problem with testing actually. i'll add a check to that

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,161 @@
+
+  private void setHttpHeaders(Request request, Response response, BinaryContent content)
+      throws MimeTypeException {
+    String mimeType = content.getMimeTypeValue();
```
can this ever return null?

Corey-Collins replied:
yes. like above, i ran into that issue with testing so i'll also add a check into that

bellcc commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,148 @@
+  private static final String QUERY_PARAM = ":transformerId";
+  private static final String RETURN_ID = "kml";
+  private static final String MIME_TYPE = "application/vnd.google-earth.kml+xml";
+  private static final String SAFE_BODY =
```
✏️ This could be defined in a `.json` file and then read in to remove this string from the codebase.

bellcc commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,148 @@
+
+    String res = (String) cqlTransformHandler.handle(mockRequest, mockResponse);
+
+    assertThat(res, is(SERVICE_NOT_FOUND));
```
❓ Can we also test the status code here too? Should be `404 Not Found` right? Also should we write different tests to track all of the possible status codes that this endpoint can produce?

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,161 @@
+
+  private String getFileExtFromMimeType(String mimeType) throws MimeTypeException {
+    MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
+    return allTypes.forName(mimeType).getExtension();
```
`forName(mimeType)` can throw `MimeTypeException` ; i think an NPE would occur in that case

jrnorth commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,150 @@
+
+  @Test
+  public void testServiceFoundWithValidResponseAndGzip() throws Exception {
+    when(mockRequest.headers(HttpHeaders.ACCEPT_ENCODING)).thenReturn(GZIP);
```
There should be more assertions in these tests. We should check that the handler is attaching the file and that it is setting the correct HTTP headers and status codes.

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlQueryResponse.java:
```diff
@@ -50,6 +51,8 @@
 
   private final Status status;
 
+  private final QueryResponse queryResponse;
```
added CqlTransformHandlerTest.java

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,175 @@
+    } catch (Exception e) {
+      LOGGER.debug("Error fetching cql request, empty or invalid body.");
+      response.status(HttpStatus.BAD_REQUEST_400);
+      return mapper.toJson(ImmutableMap.of("message", "Bad request"));
```
Do you actually need to do this? I think you can modify the request to do it automatically:

See:
https://github.com/connexta/gsr-yorktown/blob/master/catalog/aoi/src/main/java/mil/nga/gsr/catalog/aoi/application/AOIApplication.java#L103

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,175 @@
+    if (queryResponseTransformer == null) {
+      LOGGER.debug("Could not find transformer with id: {}", transformerId);
+      response.status(HttpStatus.NOT_FOUND_404);
+      return mapper.toJson(ImmutableMap.of("message", "Service not found"));
```
Same comment about the JSON. I won't duplicate this comment, but something to consider.

Corey-Collins replied:
i added mapper::toJson to transform the return of the handler

Schachte commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,175 @@
+
+    if (mimeType == null) {
+      LOGGER.debug("Failure to fetch file extension, mime-type is empty");
+      throw new MimeTypeException("Binary Content contains null mime-type value.");
```
What exceptions get thrown back to the user when we throw exceptions like this?

Corey-Collins replied:
Spark catches the exception and returns 500 with response "Error submitting feedback"

Schachte commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,207 @@
+  private CqlTransformHandler cqlTransformHandler;
+  private BinaryContent binaryContent;
+
+  @Mock private ServiceReference<QueryResponseTransformer> mockQueryResponseTransformer;
```
Do these tests actually run with @mock annotation without specifying the mockito test runner?

Corey-Collins replied:
they do

Bdthomson replied:
@Schachte You can use the `RunWith` test runner or you can use `initMocks(this)` in your setup method.

Schachte commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,207 @@
+
+  private MockResponse mockResponse;
+
+  @Before
```
❓ This might want to go into @BeforeClass since @Before will run it before every unit test and unnecessarily increase build times.

Corey-Collins replied:
beforeclass can only run if the method is static

rzwiefel commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/util/EndpointUtil.java:
```diff
@@ -392,6 +409,55 @@ public String getJson(Object result) {
     return objectMapper.toJson(result);
   }
 
+  public CqlQueryResponse executeCqlQuery(CqlRequest cqlRequest)
```
❓ Just curious what motivations for the move to EndpointUtil

Corey-Collins replied:
it is being used in both QueryApplication and CqlTransformHandler so rather than duplicate the method it was moved to a utility class

Bdthomson commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,186 @@
+    BinaryContent content =
+        bundleContext
+            .getService(queryResponseTransformer)
+            .transform(cqlQueryResponse.getQueryResponse(), Collections.emptyMap());
```
❗️ You're sending an empty map as the second parameter to this query response transformer. That means you're not sending information about Hidden Fields, Order, or Aliases.

Please see the existing CsvQueryResponse transformer to see what I mean:

/catalog/transformer/catalog-transformer-csv-queryresponse/src/main/java/ddf/catalog/transformer/csv/CsvQueryResponseTransformer.java@master#L93

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,224 @@
+
+    attachFileToResponse(request, response, queryResponseTransformer, cqlQueryResponse, arguments);
+
+    return "";
```
❓ This seems strange...the method has an `Object` return type and here we're returning and empty string. Why?

Corey-Collins replied:
@leo-sakh Whatever object is returned is mapped to json. So if there are any failures then a map is returned which will return to the ui as json. 
But in the case of success the response body is overridden with the file content. So even if "success" is sent back it will be overridden and the user won't see it.
The same is used for the CsvTransform endpoint https://github.com/codice/ddf/blob/18e0cb967af4ff722900905a20f6649629e2a0af/catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/MetacardApplication.java#L633

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,224 @@
+    String fileExt = allTypes.forName(mimeType).getExtension();
+    if (fileExt == null || StringUtils.isEmpty(fileExt)) {
+      LOGGER.debug("Fetching file extension from mime-type resulted in null or empty.");
+      throw new NullPointerException("Failure fetching file extension from mime type");
```
We shouldn't throw an NPE; we should throw some other type of checked exception (maybe IllegalArgument)

Corey-Collins replied:
@bdeining changes added

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,224 @@
+      LOGGER.debug("Fetching file extension from mime-type resulted in null or empty.");
+      throw new NullPointerException("Failure fetching file extension from mime type");
+    }
+    return allTypes.forName(mimeType).getExtension();
```
❓ Isn't this already what's in `fileExt` at line 181? Why not just return that?

Corey-Collins replied:
you're right, that should be fixed

leo-sakh commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  private Boolean containsGzip(Request request) {
+    return request.headers(HttpHeaders.ACCEPT_ENCODING).toLowerCase().contains(GZIP);
```
❓ Is it possible that there is no `ACCEPT_ENCODING` header in the request which would lead to an NPE here?

Corey-Collins replied:
there's a possibility. adding a NPE check

leo-sakh replied:
You could do something like

```
Optional.ofNullable(request.headers(HttpHeaders.ACCEPT_ENCODING)).map(String::toLowerCase).orElse("").contains(GZIP);
```
which should avoid an explicit NPE check

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,231 @@
+      throw new MimeTypeException("Binary Content contains null mime-type value.");
+    }
+
+    String fileExt = getFileExtFromMimeType(mimeType);
```
@bdeining added null check

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,236 @@
+
+    if (fileExt == null) {
+      LOGGER.debug("Invalid file extension from mimetype {}", mimeType);
+      throw new NullPointerException("Mime-type resulted in null file extension");
```
we shouldn't throw an npe

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,236 @@
+    Boolean containsGzip = false;
+    try {
+      containsGzip = request.headers(HttpHeaders.ACCEPT_ENCODING).toLowerCase().contains(GZIP);
+    } catch (NullPointerException e) {
```
why are we catching an npe here?

Corey-Collins replied:
`request.headers("something-invalid").toLowerCase().contains(GZIP)` throws NPE

bdeining replied:
should be calling `request.headers("something-invalid")` and checking if isn't null instead of catching an npe

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,236 @@
+      cqlRequest = mapper.readValue(body, CqlRequest.class);
+      if (cqlRequest == null) {
+        LOGGER.debug("Cql request parsed from body evaluated to  null.");
+        throw new NullPointerException("Cql request is null.");
```
we shouldn't throw an npe here

bdeining commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,236 @@
+
+    try {
+      arguments = mapper.readValue(body, Arguments.class).getSerializableArguments();
+    } catch (NullPointerException e) {
```
why are we catching an npe here?

bdeining commented:
build now

rzwiefel commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,233 @@
+        LOGGER.debug("Cql request parsed from body evaluated to  null.");
+        throw new IllegalArgumentException("Cql request is null.");
+      }
+    } catch (Exception e) {
```
❓ Do we have to catch `Exception` here? Can we narrow this down to the specific exceptions that could be thrown?

rzwiefel commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,233 @@
+
+  private BundleContext bundleContext;
+
+  private static final String GZIP = "gzip";
```
✏️ Its nice to group the variables together based on access modifier and attributes (eg, putting all the `private static final` things together) unless of course you specifically want it in this order for another reason.

rzwiefel commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,207 @@
+  private static final String SERVICE_NOT_FOUND = "{\"message\":\"Service not found\"}";
+  private static final String SERVICE_SUCCESS = Boon.toJson("");
+  private static final String ATTACHMENT_REGEX =
+      "^attachment;filename=export-\\d\\d\\d\\d-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d\\.\\d\\d\\dZ."
```
✏️ Couple things here
1) If you want you can specify a greedy quantifier To match something X number of times, as `{X}`. So instead of `\\d\\d\\d\\d` you can do `\\d{4}`. Check out `Greedy Quantifiers` in the [javadoc for Pattern](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html) So this could be shortened a little.
2) if the regex there was an arbitrary iso8601 time then it gets [really, really advanced to actually cover most the cases](https://stackoverflow.com/a/28022901)... but since this is just coming from an `Instant.now().toString` it will always actually be an iso8601 formatted as above. 
3) if this wasn't a test class ( I mean you still could if you want to ) I'd recommend compiling the Pattern so that it isn't recompiled every call to String::matches.

jrnorth commented on catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandlerTest.java:
```diff
@@ -0,0 +1,207 @@
+
+    assertThat(res, is(SERVICE_SUCCESS));
+    assertThat(mockResponse.status(), is(HttpStatus.OK_200));
+    assertTrue(
```
Actually we have a Maven variable for the newest version of Hamcrest so you could add this dependency instead:
```
<dependency>
  <groupId>org.hamcrest</groupId>
  <artifactId>hamcrest-junit</artifactId>
  <version>${hamcrest-junit.version}</version>
  <scope>test</scope>
</dependency>
```

then use the [`matchesPattern()`](http://hamcrest.org/JavaHamcrest/javadoc/2.0.0.0/org/hamcrest/Matchers.html#matchesPattern-java.lang.String-) matcher.

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,233 @@
+        LOGGER.debug("Cql request parsed from body evaluated to  null.");
+        throw new IllegalArgumentException("Cql request is null.");
+      }
+    } catch (Exception e) {
```
it is catching a potential Boon Json Exception or ArrayOutOfBounds (if body = ""). I'm touching up the cqlRequest validation though, I don't believe in any case cqlRequest can actually result in null, it would just have null properties

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,233 @@
+      return ImmutableMap.of("message", "Bad request");
+    }
+
+    Arguments arguments = mapper.readValue(body, Arguments.class);
```
i thought of that but i believe it would be much trickier.
certain QueryResponseTransformers have different types of arguments that can be passed to them. so to leave it arbitrary I left it to be sent as json, which is what the csv transformer similarly does.

overall I'm just not sure how it would work as url query params. 
it could be difficult to separate values if some are meant to be comma separated and others aren't. 
and the parameters can't be verbatim for instance ?hiddenvalues=x,y,z&columnorder=a,b,c
some transformers may need something different from `hidden-values` and `column-ordering`, and since transformer arguments are already read as a map it makes sense to have the json `arguments: ` element be read in the same

Corey-Collins commented:
build now

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,233 @@
+      }
+    } catch (Exception e) {
+      LOGGER.debug("Error fetching cql request");
+      response.status(HttpStatus.BAD_REQUEST_400);
```
i think the same way. but there are so many contradicting comments in this pr history about adding them i think it's best to just leave it as it is now

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,233 @@
+      throws MimeTypeException, IllegalArgumentException {
+    MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
+    String fileExt = allTypes.forName(mimeType).getExtension();
+    if (fileExt == null || StringUtils.isEmpty(fileExt)) {
```
good to know

Corey-Collins commented:
build now

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,222 @@
+    this.util = endpointUtil;
+  }
+
+  public class Arguments {
```
@jrnorth I originally tried but it doesn't work with boon

Corey-Collins commented on catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java:
```diff
@@ -0,0 +1,222 @@
+  private String getFileExtFromMimeType(String mimeType) throws MimeTypeException {
+    MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
+    String fileExt = allTypes.forName(mimeType).getExtension();
+    if (StringUtils.isEmpty(fileExt)) {
```
yeah that's a weird condition if that happens. I'll only include the debug log then

Corey-Collins commented:
build now

Corey-Collins commented:
build now

Corey-Collins commented:
build now

bdeining commented:
build now

bellcc commented:
Hero build and test successful
